### DISCUSSION TITLE
page.set_rotation() Over Rotating Some Pages

### DISCUSSION THREAD

**Discussion by MAbdElRaouf** [pypdfium2-team/pypdfium2#300] [Q&A] [answered]
### Checklist

- [X] I confirm to be using an **official package** of `pypdfium2` from `PyPI` or `GitHub/pypdfium2-team`.
- [X] I have read and acknowledged the [response policy](https://github.com/pypdfium2-team/pypdfium2#contributions).

### Description

I have a 4-page PDF where all the pages are visually rotated 270 degrees (90° counterclockwise). 

I set the rotation of all pages to 90° (`page.set_rotation(90)`) and then save to a new file (`pdf.save(path)`). **The problem is the first 2 pages are now correctly rotated (0°), while the last 2 become 90° rotated clockwise.**

Running this code to check the original pages' detected rotation and mediabox values:

```python
for page_idx in range(len(pdf_content)):

    page = pdf_content[page_idx]

    print("Page #{}:".format(page_idx), page.get_rotation(), page.get_mediabox())
```

 I found out that the first and last 2 pages have different values:
```
Page #0: 0 (0.0, 0.0, 595.2000122070312, 841.4400024414062)
Page #1: 0 (0.0, 0.0, 595.2000122070312, 841.4400024414062)
Page #2: 270 (0.0, 0.0, 841.4400024414062, 595.2000122070312)
Page #3: 270 (0.0, 0.0, 841.4400024414062, 595.2000122070312)
```

Shouldn't `page.set_rotation(90)` always rotate exactly 90° clockwise regarding of the values above? If not, what would be the correct use of `set_rotation` in my case?

Thank you.

### Install Info

```shell
pypdfium2 4.26.0
pdfium 122.0.6233.0 at D:\Python\Lib\site-packages\pypdfium2_raw\pdfium.dll

Python 3.11.5 (tags/v3.11.5:cce6ba9, Aug 24 2023, 14:38:34) [MSC v.1936 64 bit (AMD64)]

Windows-10-10.0.19045-SP0

Name: pypdfium2
Version: 4.26.0
Summary: Python bindings to PDFium
Home-page: https://github.com/pypdfium2-team/pypdfium2
Author: pypdfium2-team
Author-email: geisserml@gmail.com
License: (Apache-2.0 OR BSD-3-Clause) AND LicenseRef-PdfiumThirdParty
Location: \Python\Lib\site-packages
Requires:
```

### Validity

- [X] I confirm that I ran all commands, and pasted the whole output.

**Comment by mara004** [comment, marked as answer]
Rotation is just a programmatic value to be applied on top of the base coordinate system. It doesn't necessarily align with visuals.
`set_rotation()` quite simply sets the `/Rotation` entry in the PDF's page dictionary. In that sense, it's behaving correctly, even if, in this case, the result does not align with human visual expectations.

From what you've written I'd guess that `set_rotation(90)` for page numbers [1,2] and `set_rotation(0)` for [3,4] would result in no *visual* rotation.

I'm not aware of a straightforward way to automatically handle cases where visual and programmatic rotation don't match.
You'd probably have to do layout analysis, which is complex and requires much more computation than setting a dictionary key.

**Reply by MAbdElRaouf** [reply]
Thank you for your reply.

`set_rotation(90)` for pages [1, 2] works, for pages [3, 4] `set_rotation(0)` actually _surprisingly_ works and the pages are correctly rotated. Which is strange.

I used [PyPdf](https://github.com/py-pdf/pypdf)'s `page.rotate(90)` method on all 4 pages from the original PDF and it worked perfectly. Why can't pypdfium offer the same intuitive behavior/output?

**Reply by MAbdElRaouf** [reply]
@mara004 Please let me know if I should open a feature request instead.

**Reply by mara004** [reply]
> I used [PyPdf](https://github.com/py-pdf/pypdf)'s page.rotate(90) method on all 4 pages from the original PDF and it worked perfectly. Why can't pypdfium offer the same intuitive behavior/output?

Ah, pypdf's rotate() increments rotation. pypdfium2's set_rotation() defines the absolute rotation.
If you want the same behavior as pypdf, then add the value of get_rotation(), while keeping the final value in range >= 0 and <= 360.

I overlooked in your above report that pages 3,4 have an existing rotation of 270, so rotating by 90 would add up to 360, which is equivalent to 0.
In other words, I was too fast to assume that the problem were a mismatch between programmatic and visual state (which is also possible).

**Reply by mara004** [reply]
> set_rotation(90) for pages [1, 2] works, for pages [3, 4] set_rotation(0) actually surprisingly works and the pages are correctly rotated. Which is strange.

In that light (absolute vs relative rotation), it's not strange. 😆

**Reply by MAbdElRaouf** [reply]
> If you want the same behavior as pypdf, then add the value of get_rotation(), while keeping the final value in range >= 0 and <= 360.

Thanks, I tried this and it worked. Just needed to add a step to check if (`desired_rotation_angle + page.get_rotation()`) value is 360 and if so change it to 0, since `set_rotation()` only accepts 0, 90, 180 and 270.

Thank you for the help and the great work with pypdfium.

**Reply by mara004** [reply]
> Just needed to add a step to check if (desired_rotation_angle + page.get_rotation()) value is 360 and if so change it to 0, since set_rotation() only accepts 0, 90, 180 and 270.

Yes. But I think `% 360` (modulo) will be easier / more universal than an if-check against 360.
